pkg/transport/queuestorage/file.go

```diff
@@ -1,164 +1,328 @@
+				Str("key", f.key).
+				Str("filePath", f.path).
+				Msg("maximum file age reached, closing file")
+
+			if err := f.close(false); err != nil {
+				log.Err(err).
+					Str("key", f.key).
 					Str("filePath", f.path).
-					Msg("Maximum file age reached, closing file")
-				err := f.Close()
-				if err != nil {
-					log.Err(err).Send()
-				}
-				return
+					Msg("ticker unable to close file")
 			}
 		}
-	}
+	})
+
+	return nil
 }
 
-func (f *FileWriter) Write(data []byte) (n int, err error) {
-	f.canWrite.Lock()
-	defer f.canWrite.Unlock()
+// ensureWritable checks for constraints before any write operation.
+func (f *FileWriter) ensureWritable(dataSize int64) (err error) {
+	var openNew bool
+	defer func() {
+		if openNew {
+			err = f.close(false)
+		}
+	}()
+
+	// check to see if we will hit our row limit
+	if f.maxRows <= 0 {
+		openNew = true
+		return
+	}
 
 	// check to see if we will hit our file size limit
-	fileInfo, err := f.fd.Stat()
+	var fileInfo os.FileInfo
+	fileInfo, err = f.fd.Stat()
 	if err != nil {
 		return
 	}
 
-	if fileInfo.Size()+int64(len(data)) > f.maxFileSize {
-		if err := f.Close(); err != nil {
-			log.Err(err).Send()
-		}
-		err = fmt.Errorf("file size limit reached")
+	newSize := fileInfo.Size() + dataSize
+	if newSize > f.maxFileSize {
+		openNew = true
 		return
 	}
 
+	return
+}
+
+// postOps uploads the current file and queues its detail
+func (f *FileWriter) postOps() error {
+	log.Info().
+		Str("key", f.key).
+		Str("filePath", f.path).
+		Msg("uploading file")
+
+	file, err := os.Open(f.path)
+	if err != nil {
+		log.Error().Err(err).
+			Str("filePath", f.path).
+			Msg("unable to open file for upload")
+		return err
+	}
+
+	if err := f.storage.Upload(f.path, file); err != nil {
+		log.Error().Err(err).
+			Str("filePath", f.path).
+			Msg("unable to upload file")
+		return err
+	}
+
+	log.Info().
+		Str("key", f.key).
+		Str("filePath", f.path).
+		Msg("queuing file")
+
+	var bb []byte
+	if bb, err = json.Marshal(models.FileUploadMessage{
+		Key:  f.key,
+		Path: f.path,
+	}); err != nil {
+		log.Error().Err(err).
+			Str("key", f.key).
+			Str("filePath", f.path).
+			Msg("unable to marshal file upload message")
+		return err
+	}
+
+	if err := f.queue.Enqueue(bb); err != nil {
+		log.Error().Err(err).
+			Str("key", f.key).
+			Str("filePath", f.path).
+			Msg("unable to enqueue file")
+		return err
+	}
+
+	// TODO: Consider removing file after upload
+
+	return nil
+}
+
+// Write writes a line of data to the file. It returns the number
+// of bytes written and an error, if any. Write returns
+// a non-nil error when n != len(b) or a constraint is unmet.
+func (f *FileWriter) Write(data []byte) (n int, err error) {
+	f.mu.Lock()
+	defer f.mu.Unlock()
+
+	if f.terminated {
+		return 0, fmt.Errorf("file writer is terminated")
+	}
+
+	rowID := ulid.Make().String()
+	if data, err = sjson.SetBytes(data, "__row_id", rowID); err != nil {
+		log.Err(err).Msg("unable to set __row_id in JSON")
```

Comment:
Should this also return an error instead of just logging it?

Reply:
We want writes to succeed regardless of both attributes. An error from this call should not hinder write.